cpp/cuproj/CMakeLists.txt

```
@@ -0,0 +1,276 @@
+set(CUPROJ_BUILD_TESTS ${BUILD_TESTS})
+set(CUPROJ_BUILD_BENCHMARKS ${BUILD_BENCHMARKS})
+
+set(CUPROJ_CXX_FLAGS "")
+set(CUPROJ_CUDA_FLAGS "")
+set(CUPROJ_CXX_DEFINITIONS "")
+set(CUPROJ_CUDA_DEFINITIONS "")
+
+# Set RMM logging level
+set(RMM_LOGGING_LEVEL "INFO" CACHE STRING "Choose the logging level.")
+set_property(CACHE RMM_LOGGING_LEVEL PROPERTY STRINGS "TRACE" "DEBUG" "INFO" "WARN" "ERROR" "CRITICAL" "OFF")
+message(STATUS "CUPROJ: RMM_LOGGING_LEVEL = '${RMM_LOGGING_LEVEL}'.")
+
+###################################################################################################
+# - conda environment -----------------------------------------------------------------------------
+
+rapids_cmake_support_conda_env(conda_env MODIFY_PREFIX_PATH)
+
+###################################################################################################
+# - compiler options ------------------------------------------------------------------------------
+
+rapids_cuda_init_runtime(USE_STATIC ${CUDA_STATIC_RUNTIME})
+
+rapids_find_package(
+  CUDAToolkit REQUIRED
+  BUILD_EXPORT_SET cuproj-exports
+  INSTALL_EXPORT_SET cuproj-exports
+)
+# * find CUDAToolkit package
+# * determine GPU architectures
+# * enable the CMake CUDA language
+# * set other CUDA compilation flags
+include(cmake/modules/ConfigureCUDA.cmake)
+
+###################################################################################################
+# - dependencies ----------------------------------------------------------------------------------
+
+# add third party dependencies using CPM
+rapids_cpm_init()
+# find or add cuDF
+include(cmake/thirdparty/CUPROJ_GetCUDF.cmake)
+# find or install GoogleTest and Proj
+if (CUPROJ_BUILD_TESTS)
+    include(cmake/thirdparty/get_gtest.cmake)
+    include(cmake/thirdparty/get_proj.cmake)
+endif()
+
+###################################################################################################
+# - library targets -------------------------------------------------------------------------------
+
+add_library(cuproj
+    src/test.cu
+)
+
+set_target_properties(cuproj
+    PROPERTIES BUILD_RPATH                         "\$ORIGIN"
+               INSTALL_RPATH                       "\$ORIGIN"
+               # set target compile options
+               CXX_STANDARD                        17
+               CXX_STANDARD_REQUIRED               ON
+               CUDA_STANDARD                       17
+               CUDA_STANDARD_REQUIRED              ON
+               POSITION_INDEPENDENT_CODE           ON
+               INTERFACE_POSITION_INDEPENDENT_CODE ON
+)
+
+target_compile_options(cuproj
+            PRIVATE "$<$<COMPILE_LANGUAGE:CXX>:${CUPROJ_CXX_FLAGS}>"
+                    "$<$<COMPILE_LANGUAGE:CUDA>:${CUPROJ_CUDA_FLAGS}>"
+)
+
+target_compile_definitions(cuproj
+            PUBLIC "$<$<COMPILE_LANGUAGE:CXX>:${CUPROJ_CXX_DEFINITIONS}>"
+                   "$<BUILD_INTERFACE:$<$<COMPILE_LANGUAGE:CUDA>:${CUPROJ_CUDA_DEFINITIONS}>>"
```

Is this only desired in the build interface, or should it be public in both build and install targets?

```suggestion
                   "$<$<COMPILE_LANGUAGE:CUDA>:${CUPROJ_CUDA_DEFINITIONS}>"
```

---

Going outside of my understanding. Don't really understand cmake "interfaces", "export", "public/private".

Know of a concise explanation?

---

Like everything, this is copied from cuspatial.

---

`PUBLIC`, `PRIVATE`, and `INTERFACE` describe the transitive nature of a property through the build graph. They are used to express requirements to build and use a library.

Example:
```cmake
target_compile_definitions(cuproj 
  PRIVATE   "CUPROJ_USE_ZLIB"
  INTERFACE "CUPROJ_VERSION=1"
  PUBLIC    "CUPROJ_HAS_NVTX"
```
PRIVATE: Every source file in `cuproj` will have the `CUPROJ_USE_ZLIB` definition 
INTERFACE: Every source in consumers of `cuproj` will have the `CUPROJ_HAS_NVTX` definition 
PUBLIC: Every source file in `cuproj` and any source file of a consumer ( links to cuproj using `target_link_libraries` ) will have the `CUPROJ_HAS_NVTX` definition. ( Notice that PUBLIC is the union of PRIVATE and INTERFACE ).


So with that out of the way we can look at what `BUILD_INTERFACE` and `INSTALL_INTERFACE` means. The `INTERFACE` portion we know from above means that these rules will only impact consumers. The `BUILD` and `INSTALL` determine which consumers. The quickest way to explain it is that `BUILD` means any other consumer inside the same build, while `INSTALL` means consumers that get your project via `find_package` ( aka conda, pip, etc )

A handy guide to how to read the combo of keywords is:

- PUBLIC + BUILD_INTERFACE: Every source file in `cuproj` and any source file of a consumer inside the build directory ( aka cuproj tests )
- PUBLIC + INSTALL_INTERFACE: : Only source file of consumers post install ( conda, pip, etc ) will have this define
- INTERFACE + INSTALL_INTERFACE: Equal to PUBLIC + INSTALL_INTERFACE

Exports is an overloaded term in CMake since it maps to a poorly named command, a CMake concept, and a rapids-cmake concept. I expect you are interested in the concept of build and install export sets from rapids-cmake. The export sets in rapids-cmake calls track what project level dependencies cuproj requires ( since that currently isn't automatically done by CMake ). So when we say:

```cmake
rapids_find_package(
  CUDAToolkit REQUIRED
  BUILD_EXPORT_SET cuproj-exports
  INSTALL_EXPORT_SET cuproj-exports
)
```
It means that both install ( aka `INSTALL_INTERFACE` ) and build ( aka `BUILD_INTERFACE`)  consumers of cuproj will have need to call `find_package(CUDAToolkit)`, which rapids-cmake will do for them.

---

Thanks RobGPT! :)

So for the cuProj cxx definitions, I guess consumers need those if using its headers (right now it is header-only). Does that mean it should be public+install?

---

> So for the cuProj cxx definitions, I guess consumers need those if using its headers (right now it is header-only). Does that mean it should be public+install?

The current usage of just `PUBLIC` is correct since you want cuProj tests ( BUILD_INTERFACE ) to have those defines as well as external consumers ( INSTALL_INTERFACE )

---

Right so as I understand it currently it has `PUBLIC` + `BUILD_INTERFACE`, so I think I need to remove `BUILD_INTERFACE` as Paul suggests.

---

>so I think I need to remove BUILD_INTERFACE as Paul suggests.
Correct

---

Done.